### PR TITLE
HADOOP-16505. Add ability to register custom signer with AWS SignerFactory

### PR DESCRIPTION
**Pull request by viczsaurav**
Currently, the AWS SignerFactory restricts the class of Signer algorithms that can be used. 
We require an ability to register a custom Signer. The `SignerFactory` supports this functionality through its `registerSigner` method.

By providing a fully qualified classname to the existing parameter `fs.s3a.signing-algorithm`, the custom signer can be registered.

### CODE REVIEW

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  protected Configuration createConfiguration() {
+    Configuration conf = super.createConfiguration();
+    conf.set(Constants.SIGNING_ALGORITHM,
+            S3ATestConstants.S3A_SIGNING_ALGORITHM);
+    LOG.debug("Inside createConfiguration...");
+    return conf;
+  }
+
+  @Test
+  public void testCustomSignerOverride() throws AssertionError {
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: Please use a testing principle in your test: Arrange, Act, Assert - http://wiki.c2.com/?ArrangeActAssert

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    LOG.debug("Inside createConfiguration...");
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: These debug messages are not useful, please remove them.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    assertTrue(assertIsCustomSignerLoaded());
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: This does not need to be factored out in a separate method.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  private boolean assertIsCustomSignerLoaded() {
+    final ClientConfiguration awsConf = new ClientConfiguration();
+    LOG.debug("Inside assertIsCustomSignerLoaded...");
+    return assertEquals(awsConf.getSignerOverride(),
+            S3ATestConstants.S3A_SIGNING_ALGORITHM
+    );
+  }
+
+  private boolean assertEquals(String str1, String str2) {
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: Please don't define your own assert method if it's not needed. Use the built-in assertEquals instead from`org.junit.Assert#assertEquals(java.lang.String, java.lang.Object, java.lang.Object)`

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java (hunk 3 above)
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: If you do an assertion - like assertTrue - add your message with `org.junit.Assert#assertTrue(java.lang.String, boolean)`, so when the test faile we will know why.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 67 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1075 | trunk passed |
| +1 | compile | 31 | trunk passed |
| +1 | checkstyle | 19 | trunk passed |
| +1 | mvnsite | 35 | trunk passed |
| +1 | shadedclient | 706 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | trunk passed |
| 0 | spotbugs | 52 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 51 | trunk passed |
| -0 | patch | 70 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 35 | the patch passed |
| +1 | compile | 26 | the patch passed |
| +1 | javac | 26 | the patch passed |
| -0 | checkstyle | 18 | hadoop-tools/hadoop-aws: The patch generated 2 new + 7 unchanged - 1 fixed = 9 total (was 8) |
| +1 | mvnsite | 27 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 746 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 22 | the patch passed |
| +1 | findbugs | 59 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 70 | hadoop-aws in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3113 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1280 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux f8497e9eeb77 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9b8359b |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/1/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 41 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1447 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 39 | trunk passed |
| +1 | shadedclient | 808 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | trunk passed |
| 0 | spotbugs | 57 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 54 | trunk passed |
| -0 | patch | 76 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 64 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| -0 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 3 new + 7 unchanged - 1 fixed = 10 total (was 8) |
| +1 | mvnsite | 35 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 815 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 22 | the patch passed |
| +1 | findbugs | 71 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 70 | hadoop-aws in the patch passed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 3711 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1280 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e6e698e23c1e 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6244502 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/2/testReport/ |
| Max. process+thread count | 351 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java (hunk 3 above)
Comment (viczsaurav) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3ASigningAlgorithmOverride.java]: Refactored the code

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 37 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1250 | trunk passed |
| +1 | compile | 31 | trunk passed |
| +1 | checkstyle | 23 | trunk passed |
| +1 | mvnsite | 38 | trunk passed |
| +1 | shadedclient | 782 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | trunk passed |
| 0 | spotbugs | 57 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 55 | trunk passed |
| -0 | patch | 76 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 35 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| -0 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 3 new + 7 unchanged - 1 fixed = 10 total (was 8) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 857 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | the patch passed |
| +1 | findbugs | 74 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 70 | hadoop-aws in the patch passed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 3490 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1280 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux a376409a21ec 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ee7c261 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/3/testReport/ |
| Max. process+thread count | 340 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 130 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1293 | trunk passed |
| +1 | compile | 31 | trunk passed |
| +1 | checkstyle | 23 | trunk passed |
| +1 | mvnsite | 36 | trunk passed |
| +1 | shadedclient | 816 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | trunk passed |
| 0 | spotbugs | 61 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 59 | trunk passed |
| -0 | patch | 83 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 36 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 27 | the patch passed |
| -0 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 3 new + 7 unchanged - 1 fixed = 10 total (was 8) |
| +1 | mvnsite | 31 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 839 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 68 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 77 | hadoop-aws in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 3657 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1280 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2dd52d56ba17 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3329257 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/4/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/4/testReport/ |
| Max. process+thread count | 332 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1230 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 44 | trunk passed |
| +1 | mvnsite | 36 | trunk passed |
| +1 | shadedclient | 807 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | trunk passed |
| 0 | spotbugs | 63 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 61 | trunk passed |
| -0 | patch | 83 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| -0 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 3 new + 7 unchanged - 1 fixed = 10 total (was 8) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 854 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 92 | hadoop-aws in the patch passed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 3510 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1280 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e49c3ef3ed52 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 915cbc9 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/5/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/5/testReport/ |
| Max. process+thread count | 340 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1280/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
closing this as its been superceded by the other patches from Sid

Can you experiment with the current trunk branch and verify it meets your needs? If it doesn't -now is the time to make the appropriate changes

**Pull request closed without merging**